tests/unit/moduleapi/customauth.tcl

```diff
@@ -0,0 +1,398 @@
+        catch {r exec} e
+        assert_match {*ERR Blocking module command called from transaction*} $e
+        assert_match {*calls=2,*,rejected_calls=0,failed_calls=1} [cmdstat auth]
+    }
+
+    test {Disabling Redis User during blocking custom auth} {
+        r config resetstat
+        r acl setuser foo >pwd on ~* &* +@all
+        set rd [redis_deferring_client]
+
+        # Attempt blocking custom auth and disable the Redis user while custom auth is in progress.
+        $rd AUTH foo pwd
+        wait_for_blocked_clients_count 1
+        r acl setuser foo >pwd off ~* &* +@all
+
+        # Validate that custom auth failed.
+        wait_for_blocked_clients_count 0 500 10
+        $rd flush
+        catch { $rd read } e
+        assert_match {*WRONGPASS*} $e
+        assert_match {*calls=1,*,rejected_calls=0,failed_calls=1} [cmdstat auth]
+    }
+
+    test {Killing a client in the middle of blocking custom auth} {
+        r config resetstat
+        r acl setuser foo >pwd on ~* &* +@all
+        set rd [redis_deferring_client]
+        $rd client id
+        set cid [$rd read]
+
+        # Attempt blocking custom auth command on client `cid` and kill the client while custom auth
+        # is in progress.
+        $rd AUTH foo pwd
+        wait_for_blocked_clients_count 1
+        r client kill id $cid
+
+        # Validate that the blocked client count goes to 0 and no AUTH command is tracked.
+        wait_for_blocked_clients_count 0 500 10
+        $rd flush
+        catch { $rd read } e
+        assert_match {I/O error reading reply} $e
+        assert_match {} [cmdstat auth]
+    }
+
+    test {test RM_AbortBlock Module API during blocking custom auth} {
+        r config resetstat
+        r acl setuser foo >pwd on ~* &* +@all
+
+        # Attempt custom auth. With the "block_abort" as the password, the "customauth.so" module
+        # blocks the client and uses the RM_AbortBlock API. This should result in custom auth
+        # failing and the client being unblocked with the default AUTH err message.
+        catch { r AUTH foo block_abort } e
+        assert_match {*WRONGPASS*} $e
+        assert_match {*calls=1,*,rejected_calls=0,failed_calls=1} [cmdstat auth]
+    }
+
+    test {test RM_RegisterCustomAuthCallback Module API during blocking custom auth} {
+        r config resetstat
+        r acl setuser foo >defaultpwd on ~* &* +@all
+        set rd [redis_deferring_client]
+
+        # Start the custom auth attempt with the standard Redis auth password for the user. This
+        # will result in all custom auth cbs attempted and then standard Redis auth will be tried.
+        $rd AUTH foo defaultpwd
+        wait_for_blocked_clients_count 1
+
+        # Validate that we allow modules to register custom auth cbs while custom auth is already
+        # in progress.
+        assert_equal {OK} [r testmoduletwo.rm_register_auth_cb]
+        assert_equal {OK} [r testmoduleone.rm_register_blocking_auth_cb]
```

Comment:
Yes, that is a good point. We can prevent both load and unload operations while module based authentication is going on on any client to prevent this from happening.

Reply:
Addressed here: https://github.com/redis/redis/pull/11659/commits/41043bbc94a0fc254a33faf27ee13f2300fb2818

Instead of preventing loading during custom auth, we are just unregistering the custom auth callbacks before calling `moduleUnload`